guppylang/checker/expr_checker.py

```diff
@@ -470,11 +565,46 @@ def visit_Call(self, node: ast.Call) -> tuple[ast.expr, Type]:
             if isinstance(defn, CallableDef):
                 return defn.synthesize_call(node.args, node, self.ctx)
 
-        # Otherwise, it must be a function as a higher-order value
+        # Otherwise, it must be a function as a higher-order value, or a tensor
         if isinstance(ty, FunctionType):
             args, return_ty, inst = synthesize_call(ty, node.args, node, self.ctx)
             node.func = instantiate_poly(node.func, ty, inst)
             return with_loc(node, LocalCall(func=node.func, args=args)), return_ty
+        elif (
+            isinstance(ty, TupleType)
+            and parse_function_tensor(ty)
+            and isinstance(node.func, ast.Tuple)
+        ):
+            # Note: None of the function types in a tuple of functions will have
+            # overlapping type arguments.
+            function_elems = parse_function_tensor(ty)
+            assert isinstance(function_elems, list)
+            func_ty = function_tensor_signature(function_elems)
+            remaining_args = node.args
+            return_tys: list[Type] = []
+            processed_args: list[ast.expr] = []
+            call_nodes: list[ast.expr] = []
+            for func in node.func.elts:
+                args, return_ty, inst, remaining_args = synthesize_call_with_leftovers(
+                    func_ty, remaining_args, node, self.ctx
+                )
+                processed_args.extend(args)
+                if isinstance(return_ty, TupleType):
+                    return_tys.extend(return_ty.element_types)
+                else:
+                    return_tys.append(return_ty)
+
+                if isinstance(func, GlobalName):
+                    assert isinstance(self.ctx.globals[func.def_id], CallableDef)
+                    call_nodes.append(
+                        GlobalCall(def_id=func.def_id, args=args, type_args=inst)
+                    )
```

Comment:
Similar to the checking part, you need `self.ctx.globals[func.def_id].synthesize_call()` and replace the assert with a user error